opt: Add Update support to CBO

andy-kimball opened the pull request:
Add a new Update operator to the CBO which is used to optimize the
UPDATE SQL statement. As part of compiling the UPDATE statement, the
optbuilder constructs an expression that outputs the existing values
for all rows from the target table that match the WHERE clause.
Additional column(s) that provide updated values are projected for
each of the SET expressions, as well as for any computed columns. For
example:
```
  CREATE TABLE abc (a INT PRIMARY KEY, b INT, c INT)
  UPDATE abc SET b=1 WHERE a=2
```
This would create an input expression similar to this SQL:
```
  SELECT a AS oa, b AS ob, c AS oc, 1 AS nb FROM abc WHERE a=2
```
The execution engine evaluates this relational expression and uses
the resulting values to form the KV keys and values.

andy-kimball committed:
opt: Move insertBuilder to new file

Prepare to rename to mutationBuilder and generalize it for other
mutation operators.

Release note: None

andy-kimball committed:
opt: Re-format mutation operator output

Create a separate treeprinter child for each column changed by a mutation,
like this:

  ├── insert:
  │    ├──  column1:7(int) => a:1(int)
  │    └──  column2:8(int) => b:2(int)

This is easier to read, and it will work better for Update mutations.

Release note: None

andy-kimball committed:
opt: Generalize mutationBuilder

Make mutationBuilder methods more flexible, in preparation for adding
Update/Upsert support.

Release note: None

andy-kimball committed:
opt: Rename insertBuilder to mutationBuilder

This is in preparation for adding Update/Upsert support.

Release note: None

andy-kimball committed:
opt: Don't qualify names that are identical

If two columns with the same name are added to query metadata, then
the formatter will qualify the names with the table name. This commit
changes the algorithm to not qualify if the table names are the same.
There is no extra clarity if that's the case, and it looks cluttered.

This is important for supporting the Update operator, where the same
table is always used twice in the same query.

Release note: None

andy-kimball committed:
opt: Change mutation column metadata interface

After working on Update, I decided that mutation columns should be part
of the main Column list exposed by the opt.Table interface. There are
problems with each choice, but after trying both ways, I found that most
code that calls the Column method wants to include mutation columns if
present. In the cases where that's not true, there's a new
opt.IsMutationColumn helper method to make to make it easy to exclude
them.

In addition, the new design exposes a MutationColumn struct that indicates
whether the mutation column is delete-only or write-only. This will be
needed for Update. That in turn means that Insert may not insert values
into all table columns (i.e. it won't insert into delete-only columns),
and so the ConstructInsert exec factory method takes a new insertCols
parameter that specifies the subset of table columns into which values
will be inserted.

Release note: None

andy-kimball committed:
sql: Do not automatically add mutation columns

If "wantedColumns" are specified as an argument to the scanNode, don't
add mutation columns automatically. Require caller to add them if they
want them. Note that if wantedColumns are not specified, then mutation
columns are always added (as are all other columns).

This change is important for implementing the CBO Update operator, since
it sometimes needs to fetch a subset of the mutation columns.

Release note: None

andy-kimball committed:
opt: Add Update support to CBO

Add a new Update operator to the CBO which is used to optimize the
UPDATE SQL statement. As part of compiling the UPDATE statement, the
optbuilder constructs an expression that outputs the existing values
for all rows from the target table that match the WHERE clause.
Additional column(s) that provide updated values are projected for
each of the SET expressions, as well as for any computed columns. For
example:

  CREATE TABLE abc (a INT PRIMARY KEY, b INT, c INT)
  UPDATE abc SET b=1 WHERE a=2

This would create an input expression similar to this SQL:

  SELECT a AS oa, b AS ob, c AS oc, 1 AS nb FROM abc WHERE a=2

The execution engine evaluates this relational expression and uses
the resulting values to form the KV keys and values.

Release note: None

andy-kimball committed:
opt: Add execbuilder support for the Update operator

Build an updateNode from the optimizer's new Update operator. For now,
always fetch all columns from the table. In the future, do "needed
column" analysis to select only those columns that are needed for the
update.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/32774)
<!-- Reviewable:end -->

andy-kimball commented:
Note that I commented out the `Max1Row` operator when compiling SET subqueries. That won't work properly until this PR merges:

https://github.com/cockroachdb/cockroach/pull/32753

I'll rebase once that happens. In the meantime, I want to get this PR posted, since there's a lot to review.

justinj reviewed (commented):
Only nits from me for now - might take a more holistic look in a while though

> Reviewed 2 of 2 files at r1, 2 of 2 files at r2, 6 of 6 files at r3, 10 of 10 files at r4, 33 of 33 files at r5, 20 of 20 files at r6, 3 of 3 files at r7, 21 of 21 files at r8, 21 of 21 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-:-LSq6sT69yJn8VIPUhJp:b81w7u3)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/opt_exec_factory.go, line 1009 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSq4QSb9Y6rLyLmWyoD:-LSq4QSb9Y6rLyLmWyoE:b-ek4l0a) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebc37203b4c925f8e0d22d907c786512927d397d/pkg/sql/opt_exec_factory.go#L1009)):*
> ```Go
> 	}
> 
> 	// serialize the data-modifying plan to ensure that no data is
> ```

nit: `Serialize`
___
*[pkg/sql/logictest/testdata/logic_test/default, line 111 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSq4zzo7WdQnNGl2kxS:-LSq4zzp13zczVE8OT7v:bz2rvoy) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebc37203b4c925f8e0d22d907c786512927d397d/pkg/sql/logictest/testdata/logic_test/default#L111)):*
> ```
> 
> statement ok
> UPDATE v SET b = DEFAULT
> ```

Is this no longer legal? Postgres allows it despite it being kind of silly and pointless (I don't think it's a big deal if we don't support it).
___
*[pkg/sql/opt/catalog.go, line 114 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSpFDSX6Twvk6XFhrH9:-LSpFDSY47vQ2fLwb7gQ:bfgbo46) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c514560fbc722793c30710b7892083850bea057/pkg/sql/opt/catalog.go#L114)):*
> ```Go
> 	//
> 	// If the column is not a mutation column, then ok=false.
> 	MutationColumn(i int) (col MutationColumn, ok bool)
> ```

OOC, how come this and not making `MutationColumn` and interface and using a cast to determine if a column is a mutation column or not?
___
*[pkg/sql/opt/norm/testdata/rules/decorrelate, line 86 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSppFtO0Bj_JB0nMDHf:-LSppFtO0Bj_JB0nMDHg:bhaib2) ([raw file](https://github.com/cockroachdb/cockroach/blob/382b1dcab6fffe9e1853435f23dd0a379b0aecb3/pkg/sql/opt/norm/testdata/rules/decorrelate#L86)):*
> ```
> 
> opt expect=DecorrelateJoin
> UPDATE xy SET (x, y)=(SELECT * FROM uv WHERE u=x)
> ```

This is pretty cool :thumbsup:
___
*[pkg/sql/opt/ops/statement.opt, line 32 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSon2D0B7WhUUe-iq8N:-LSon2D0B7WhUUe-iq8O:b-9vsw41) ([raw file](https://github.com/cockroachdb/cockroach/blob/7596da7c9fe4f5604ef3eb77bc8fd3428479b5d1/pkg/sql/opt/ops/statement.opt#L32)):*
> ```
> [Private]
> define MutationPrivate {
>     # Table identifies the table into which to insert. It is an id that can be
> ```

nit: `the table which is being mutated`
___
*[pkg/sql/opt/ops/statement.opt, line 60 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSpp_iy1-9z2cp2fqh6:-LSpp_iy1-9z2cp2fqh7:blkjn0m) ([raw file](https://github.com/cockroachdb/cockroach/blob/382b1dcab6fffe9e1853435f23dd0a379b0aecb3/pkg/sql/opt/ops/statement.opt#L60)):*
> ```
>     # Fetch columns are referenced by update, computed, and constraint
>     # expressions. They're also needed to formulate the final key/value pairs;
>     # updating even one column in a family requires the entire the entire value
> ```

nit: the entire the entire
___
*[pkg/sql/opt/optbuilder/scope_column.go, line 41 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSpYf5HCjB0nVODdz1r:-LSpYf5HCjB0nVODdz1s:barhvax) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c514560fbc722793c30710b7892083850bea057/pkg/sql/opt/optbuilder/scope_column.go#L41)):*
> ```Go
> 
> 	// hidden is true if the column is not selected by a '*' wildcard operator.
> 	// The column must be explicitly referenced by name, or otherwise is hidden.
> ```

nit: maybe "or otherwise is not included"? since `hidden` is what's already being discussed.
___
*[pkg/sql/opt/optbuilder/scope_column.go, line 44 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSpYlMgDea6VyGDemWo:-LSpYlMgDea6VyGDemWp:b-at74b1) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c514560fbc722793c30710b7892083850bea057/pkg/sql/opt/optbuilder/scope_column.go#L44)):*
> ```Go
> 	hidden bool
> 
> 	// mutation is true if the column is in process of being dropped or added to
> ```

nit: is in *the* process


<!-- Sent from Reviewable.io -->

rytaft reviewed (changes requested):
Still have a bit more to review, but looks good so far!

> Reviewed 2 of 2 files at r1, 2 of 2 files at r2, 6 of 6 files at r3, 10 of 10 files at r4, 33 of 33 files at r5, 20 of 20 files at r6, 3 of 3 files at r7, 14 of 21 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-:-LSqI5cq3r1XWPy3Covn:b-quwcaf)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/opt_exec_factory.go, line 981 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSq1XN7ACMSvycXv_qJ:-LSq1XN7ACMSvycXv_qK:b-si710d) ([raw file](https://github.com/cockroachdb/cockroach/blob/d80ed57516a6701d1798bdba79a8d06c3ca92df3/pkg/sql/opt_exec_factory.go#L981)):*
> ```Go
> 	colCfg := scanColumnsConfig{
> 		wantedColumns: make([]tree.ColumnID, 0, cols.Len()),
> 		visibility:    publicAndNonPublicColumns,
> ```

Should this always be the visibility? What if this is a read only query?
___
*[pkg/sql/opt/metadata.go, line 329 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSppMEyB0Jtb0FXNcCv:-LSppMEyB0Jtb0FXNcCw:bosjvee) ([raw file](https://github.com/cockroachdb/cockroach/blob/b388b5d98ab73eb63a6e8440fac5be310517e0ef/pkg/sql/opt/metadata.go#L329)):*
> ```Go
> 					}
> 				}
> 				break
> ```

Seems like you could be excluding some ambiguous cases here. What if there is one column that has the same name and is from the same table, but another column that has the same name and is from a different table? Seems like you might miss that other column.
___
*[pkg/sql/opt/exec/factory.go, line 233 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSpteyh7zQcnIAn4nIG:-LSpteyh7zQcnIAn4nIH:b16kyl5) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c514560fbc722793c30710b7892083850bea057/pkg/sql/opt/exec/factory.go#L233)):*
> ```Go
> 	// input columns are inserted into a subset of columns in the table, in the
> 	// same order they're defined. The insertCols set contains the ordinal
> 	// positions of columns in the table into which values are inserted. The
> ```

What happens to the other columns? Do they get a default value?
___
*[pkg/sql/opt/memo/check_expr.go, line 220 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSq9Dm8C6HPw-JYguCs:-LSq9Dm8C6HPw-JYguCt:bt3o5om) ([raw file](https://github.com/cockroachdb/cockroach/blob/382b1dcab6fffe9e1853435f23dd0a379b0aecb3/pkg/sql/opt/memo/check_expr.go#L220)):*
> ```Go
> 		_, ok := tab.MutationColumn(i)
> 		if ok {
> 			mutCols.Add(int(private.InsertCols[i]))
> ```

Isn't this going to panic for updates?
___
*[pkg/sql/opt/memo/statistics_builder.go, line 1752 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSqAphr8bDXTS69gLIa:-LSqAphr8bDXTS69gLIb:b-h2p5jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/382b1dcab6fffe9e1853435f23dd0a379b0aecb3/pkg/sql/opt/memo/statistics_builder.go#L1752)):*
> ```Go
> 
> func (sb *statisticsBuilder) colStatMutation(
> 	colSet opt.ColSet, insUpd RelExpr,
> ```

[nit] insUpd -> mutation (to be consistent with buildMutation)
___
*[pkg/sql/opt/memo/testdata/logprops/insert, line 39 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSp_KmH4mlHSnO-HJGc:-LSp_KmH4mlHSnO-HJGd:bpgxpvr) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcd3edce0590615727acb5f6a4b44591e1699688/pkg/sql/opt/memo/testdata/logprops/insert#L39)):*
> ```
> insert abcde
>  ├── columns: <none>
>  ├── insert:
> ```

[nit] it's a little weird to have this child node also called insert. Not sure what a better name would be, but maybe something that evokes column mappings...
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 64 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSqF4Eo9VdRAyLEYMqQ:-LSqF4Eo9VdRAyLEYMqR:b-6nkhew) ([raw file](https://github.com/cockroachdb/cockroach/blob/382b1dcab6fffe9e1853435f23dd0a379b0aecb3/pkg/sql/opt/optbuilder/mutation_builder.go#L64)):*
> ```Go
> 	// in the target table, including mutation columns. Table columns which will
> 	// not have values inserted are set to zero (e.g. delete-only mutation
> 	// columns). insertColList empty if this is not an Insert operator.
> ```

[nit] insertColList is empty
___
*[pkg/sql/opt/optbuilder/select.go, line 280 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSpzq1t5QAoaT_wL_r0:-LSpzq1uBIbx6GGTKrV6:b6ca6o5) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c514560fbc722793c30710b7892083850bea057/pkg/sql/opt/optbuilder/select.go#L280)):*
> ```Go
> 	ordinals []int,
> 	indexFlags *tree.IndexFlags,
> 	includeMutations bool,
> ```

This is shadowing the constants you defined above - was that intentional?


<!-- Sent from Reviewable.io -->

rytaft reviewed (changes requested):
Nice work! Done with the first pass.

> Reviewed 7 of 21 files at r8, 21 of 21 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-:-LSuaXAFAuKMMQsyOBx-:b-n592bw)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/opt_exec_factory.go, line 974 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSuXNy0-8ab4JbvmUIc:-LSuXNy0-8ab4JbvmUId:b-awp3it) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebc37203b4c925f8e0d22d907c786512927d397d/pkg/sql/opt_exec_factory.go#L974)):*
> ```Go
> 	}
> 
> 	// Determine the relational type of the generated insert node.
> ```

update node?
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 609 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSu4NLU9GsLuROGlKgv:-LSu3ds72kT_lhl2-LC4:b-djpbk7) ([raw file](https://github.com/cockroachdb/cockroach/blob/382b1dcab6fffe9e1853435f23dd0a379b0aecb3/pkg/sql/opt/optbuilder/mutation_builder.go#L609)):*
> ```Go
> 		// references (like computed expressions) should point to the new column
> 		// rather than the old.
> 		projectionsScope.cols[ord].name = ""
> ```

This is a bit confusing. It might help to say that `sourceCol` was added to the `projectionsScope` as well (or will be added later). Would also help to explain somewhere why the table columns are in the same scope at all (e.g., why not just add a new scope with the old columns replaced).
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 655 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSu5ffA29jFJoFGGgxL:-LSu5YpA6Fq3QNx3Ajdr:b-oyyz79) ([raw file](https://github.com/cockroachdb/cockroach/blob/382b1dcab6fffe9e1853435f23dd0a379b0aecb3/pkg/sql/opt/optbuilder/mutation_builder.go#L655)):*
> ```Go
> 				mb.outScope.expr = mb.b.factory.ConstructLeftJoinApply(
> 					mb.outScope.expr,
> 					//mb.b.factory.ConstructMax1Row(subqueryScope.expr),
> ```

Don't forget to uncomment (I presume this depends on the Max1Row issue...)
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 708 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSu7G4D81PdfGUBp3NQ:-LSu7G4D81PdfGUBp3NR:b497flf) ([raw file](https://github.com/cockroachdb/cockroach/blob/382b1dcab6fffe9e1853435f23dd0a379b0aecb3/pkg/sql/opt/optbuilder/mutation_builder.go#L708)):*
> ```Go
> 	// Allow mutation columns to be referenced by other computed mutation columns
> 	// (otherwise the scope will raise an error if a mutation column is
> 	// referenced).
> ```

Do we need to reset the mutation field later?
___
*[pkg/sql/opt/optbuilder/testdata/update, line 1108 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSuSgJj99vMSyLWNqxy:-LSuSgJk-e6yADeDdTiW:b-4olf72) ([raw file](https://github.com/cockroachdb/cockroach/blob/382b1dcab6fffe9e1853435f23dd0a379b0aecb3/pkg/sql/opt/optbuilder/testdata/update#L1108)):*
> ```
> UPDATE mutation SET m=1 RETURNING "o:write-only"
> ----
> error (42703): column "o:write-only" does not exist
> ```

should this return the same error as below? (o:write-only is being backfilled)?


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
Great stuff! Only have a few comments so far, will give it another pass

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-:-LSv1Cjf600obXc8vR8H:b-liqcvf)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/logictest/testdata/logic_test/insert, line 652 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSuTsdV4jOyhr3UV3RB:-LSuTsdW4SjnIjrL7dMe:b-zidqhq) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c514560fbc722793c30710b7892083850bea057/pkg/sql/logictest/testdata/logic_test/insert#L652)):*
> ```
> BEGIN; ALTER TABLE mutation DROP COLUMN y
> 
> # TODO(andyk): re-enable once issue #32759 is fixed.
> ```

I think this is fixed now.
___
*[pkg/sql/logictest/testdata/logic_test/subquery, line 242 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSv0o4lBioC3Eyjtopv:-LSv0o4lBioC3Eyjtopw:bb0fi81) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebc37203b4c925f8e0d22d907c786512927d397d/pkg/sql/logictest/testdata/logic_test/subquery#L242)):*
> ```
> # Error is different between HP and CBO, but is reasonable in both cases. Just
> # verify there's an error that contains the word "columns".
> statement error columns
> ```

Nit: I think this is a regexp so it could contain both errors with `|`
___
*[pkg/sql/opt/optbuilder/scope.go, line 583 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSuV5Kx42AIY_XxbZck:-LSuV5Kx42AIY_XxbZcl:b-k9ju9p) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c514560fbc722793c30710b7892083850bea057/pkg/sql/opt/optbuilder/scope.go#L583)):*
> ```Go
> 			}
> 
> 			if err := checkNoMutationColumn(col); err != nil {
> ```

In this loop, we are looking for candidates.. Shouldn't we ignore this column if it's a mutation instead of erroring out right away? The col name could refer to a column in a parent scope. (I'd add a test like this)


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32774)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/opt_exec_factory.go, line 981 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSq1XN7ACMSvycXv_qJ:-LSu5kz07bG1EOMVksw7:b-9c3zfq) ([raw file](https://github.com/cockroachdb/cockroach/blob/d80ed57516a6701d1798bdba79a8d06c3ca92df3/pkg/sql/opt_exec_factory.go#L981)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Should this always be the visibility? What if this is a read only query?
</blockquote></details>

I added this comment:
```
	// Set visibility=publicAndNonPublicColumns, since all columns in the "cols"
	// set should be projected, regardless of whether they're public or non-
	// public. The caller decides which columns to include (or not include). Note
	// that when wantedColumns is non-empty, the visibility flag will never
	// trigger the addition of more columns.
```
___
*[pkg/sql/opt_exec_factory.go, line 974 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSuXNy0-8ab4JbvmUIc:-LSzIXc6BDnV2PZF3Dyw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebc37203b4c925f8e0d22d907c786512927d397d/pkg/sql/opt_exec_factory.go#L974)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

update node?
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/subquery, line 242 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSv0o4lBioC3Eyjtopv:-LSzLCSF8iAJU4wY4jC7:b5bychb) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebc37203b4c925f8e0d22d907c786512927d397d/pkg/sql/logictest/testdata/logic_test/subquery#L242)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Nit: I think this is a regexp so it could contain both errors with `|`
</blockquote></details>

Oh, good to know.
___
*[pkg/sql/opt/catalog.go, line 114 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSpFDSX6Twvk6XFhrH9:-LSuAO2AA1injd924amK:burl39x) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c514560fbc722793c30710b7892083850bea057/pkg/sql/opt/catalog.go#L114)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I had a similar question (can't we add methods to `Column` to retrieve this info?), but I think it's harder to implement because we use `*ColumnDescriptor` as `opt.Column`; we would need to add a wrapper (which means more allocations).
</blockquote></details>

Yes, extra allocations were one of my worries as well. But I think it's fine in this case, since mutation columns are rare. I went ahead and switched the implementation to use a wrapper.
___
*[pkg/sql/opt/metadata.go, line 329 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSppMEyB0Jtb0FXNcCv:-LSu_klI3Sp1mHsbQrT4:b-if9phx) ([raw file](https://github.com/cockroachdb/cockroach/blob/b388b5d98ab73eb63a6e8440fac5be310517e0ef/pkg/sql/opt/metadata.go#L329)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Seems like you could be excluding some ambiguous cases here. What if there is one column that has the same name and is from the same table, but another column that has the same name and is from a different table? Seems like you might miss that other column.
</blockquote></details>

I removed the `break` to avoid that case.
___
*[pkg/sql/opt/exec/factory.go, line 233 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSpteyh7zQcnIAn4nIG:-LSunE-t8zCpY2IYb2CE:bjxk7pp) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c514560fbc722793c30710b7892083850bea057/pkg/sql/opt/exec/factory.go#L233)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

What happens to the other columns? Do they get a default value?
</blockquote></details>

I added this to the comment:
```
All columns are expected to be present except delete-only mutation columns, since those do not need to participate in an insert operation.
```
___
*[pkg/sql/opt/memo/check_expr.go, line 220 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSq9Dm8C6HPw-JYguCs:-LSuiuG9E6DkschHlijG:b-9lxycj) ([raw file](https://github.com/cockroachdb/cockroach/blob/382b1dcab6fffe9e1853435f23dd0a379b0aecb3/pkg/sql/opt/memo/check_expr.go#L220)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Isn't this going to panic for updates?
</blockquote></details>

Yes, it will. Fixed.
___
*[pkg/sql/opt/memo/statistics_builder.go, line 1752 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSqAphr8bDXTS69gLIa:-LSunN4vCNI_po72rtwO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/382b1dcab6fffe9e1853435f23dd0a379b0aecb3/pkg/sql/opt/memo/statistics_builder.go#L1752)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] insUpd -> mutation (to be consistent with buildMutation)
</blockquote></details>

Done.
___
*[pkg/sql/opt/memo/testdata/logprops/insert, line 39 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSp_KmH4mlHSnO-HJGc:-LSuyDp72x8bUoJ4lpam:bfoa9i2) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcd3edce0590615727acb5f6a4b44591e1699688/pkg/sql/opt/memo/testdata/logprops/insert#L39)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] it's a little weird to have this child node also called insert. Not sure what a better name would be, but maybe something that evokes column mappings...
</blockquote></details>

I changed it to be `insert-mapping` and `update-mapping`.
___
*[pkg/sql/opt/ops/statement.opt, line 60 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSpp_iy1-9z2cp2fqh6:-LSuynrI7iH5yu7p60ov:b-sndiwu) ([raw file](https://github.com/cockroachdb/cockroach/blob/382b1dcab6fffe9e1853435f23dd0a379b0aecb3/pkg/sql/opt/ops/statement.opt#L60)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

nit: the entire the entire
</blockquote></details>

The repetition is a clever literary device that emphasizes the fact that not just one, but multiple columns may need to be updated at once.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 64 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSqF4Eo9VdRAyLEYMqQ:-LSv-Dp_0P_D2yQz3W0t:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/382b1dcab6fffe9e1853435f23dd0a379b0aecb3/pkg/sql/opt/optbuilder/mutation_builder.go#L64)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] insertColList is empty
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 609 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSu4NLU9GsLuROGlKgv:-LSv0TO_3tsjdwo4mhky:b-88pil6) ([raw file](https://github.com/cockroachdb/cockroach/blob/382b1dcab6fffe9e1853435f23dd0a379b0aecb3/pkg/sql/opt/optbuilder/mutation_builder.go#L609)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

This is a bit confusing. It might help to say that `sourceCol` was added to the `projectionsScope` as well (or will be added later). Would also help to explain somewhere why the table columns are in the same scope at all (e.g., why not just add a new scope with the old columns replaced).
</blockquote></details>

I updated the comment to be this:
```
		// "Shadow" the existing column of the same name, since any future
		// references (like computed expressions) should point to the new column
		// containing the updated value rather than the old column containing the
		// original value. The old columns need to be retained in the projection
		// because some original values are needed to formulate the update keys.
```
Is that clear enough?
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 655 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSu5ffA29jFJoFGGgxL:-LSzPgqBAdOfADDsahQl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/382b1dcab6fffe9e1853435f23dd0a379b0aecb3/pkg/sql/opt/optbuilder/mutation_builder.go#L655)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Don't forget to uncomment (I presume this depends on the Max1Row issue...)
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 708 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSu7G4D81PdfGUBp3NQ:-LSv0k0RAqUAAVq0NdJU:b6d3zu5) ([raw file](https://github.com/cockroachdb/cockroach/blob/382b1dcab6fffe9e1853435f23dd0a379b0aecb3/pkg/sql/opt/optbuilder/mutation_builder.go#L708)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Do we need to reset the mutation field later?
</blockquote></details>

No, b/c mutation columns are not projected by the Update operator. I added that to the comment.
___
*[pkg/sql/opt/optbuilder/scope.go, line 583 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSuV5Kx42AIY_XxbZck:-LSzPyne8jEsHa6wA_48:b-k2rwtz) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c514560fbc722793c30710b7892083850bea057/pkg/sql/opt/optbuilder/scope.go#L583)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

In this loop, we are looking for candidates.. Shouldn't we ignore this column if it's a mutation instead of erroring out right away? The col name could refer to a column in a parent scope. (I'd add a test like this)
</blockquote></details>

Good catch. Fixed.
___
*[pkg/sql/opt/optbuilder/select.go, line 280 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSpzq1t5QAoaT_wL_r0:-LSzI0vmEzCnq0awFNY7:b-cu0sl5) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c514560fbc722793c30710b7892083850bea057/pkg/sql/opt/optbuilder/select.go#L280)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

This is shadowing the constants you defined above - was that intentional?
</blockquote></details>

Not intentional, I fixed it.
___
*[pkg/sql/opt/optbuilder/testdata/update, line 1108 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSuSgJj99vMSyLWNqxy:-LSzI3aY3cA4wiHGBD7y:b5mclrw) ([raw file](https://github.com/cockroachdb/cockroach/blob/382b1dcab6fffe9e1853435f23dd0a379b0aecb3/pkg/sql/opt/optbuilder/testdata/update#L1108)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

should this return the same error as below? (o:write-only is being backfilled)?
</blockquote></details>

This is the error given by HP. It'd be messy to make it be the backfilled error, b/c the `RETURNING` clause does not project mutation columns, and it'd be messy to make it do so.
___
*[pkg/sql/logictest/testdata/logic_test/default, line 111 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSq4zzo7WdQnNGl2kxS:-LSuAFDe3AApJjA-nsnH:bkhsik2) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebc37203b4c925f8e0d22d907c786512927d397d/pkg/sql/logictest/testdata/logic_test/default#L111)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Is this no longer legal? Postgres allows it despite it being kind of silly and pointless (I don't think it's a big deal if we don't support it).
</blockquote></details>

I changed code to allow it, so now a NOT-NULL error would only be raised at runtime if there is at least one row to update.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 3 of 86 files at r10, 3 of 10 files at r13, 27 of 30 files at r14, 5 of 20 files at r15, 2 of 3 files at r16, 25 of 25 files at r17, 23 of 23 files at r18.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-:-LT3Q43f7KuVtIXdUdwb:b-uwkowd)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 609 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSu4NLU9GsLuROGlKgv:-LT3MsvC2cFAeXbd4qSt:b-euvz2t) ([raw file](https://github.com/cockroachdb/cockroach/blob/382b1dcab6fffe9e1853435f23dd0a379b0aecb3/pkg/sql/opt/optbuilder/mutation_builder.go#L609)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I updated the comment to be this:
```
		// "Shadow" the existing column of the same name, since any future
		// references (like computed expressions) should point to the new column
		// containing the updated value rather than the old column containing the
		// original value. The old columns need to be retained in the projection
		// because some original values are needed to formulate the update keys.
```
Is that clear enough?
</blockquote></details>

:thumbsup:  Thanks!
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 703 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LT3N7HE0vZviYTpB0Oj:-LT3N7HE0vZviYTpB0Ok:bcgsnlf) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d64f2742554bb1b5fe0492ef2efe19f3239af9b/pkg/sql/opt/optbuilder/mutation_builder.go#L703)):*
> ```Go
> 	// Allow mutation columns to be referenced by other computed mutation columns
> 	// (otherwise the scope will raise an error if a mutation column is
> 	// referenced). These does not need to be set back to true again because
> ```

[nit] These does -> These do
___
*[pkg/sql/opt/optbuilder/testdata/insert, line 730 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LT3O70a77FQHm2AKp85:-LT3O70a77FQHm2AKp86:bet8e47) ([raw file](https://github.com/cockroachdb/cockroach/blob/382b1dcab6fffe9e1853435f23dd0a379b0aecb3/pkg/sql/opt/optbuilder/testdata/insert#L730)):*
> ```
> error (42703): column "unk" does not exist
> 
> # Cannot insert null into non-null column.
> ```

Why did you take out this test?
___
*[pkg/sql/opt/optbuilder/testdata/insert, line 689 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LT3ORf0CMM3g4s-ZKg9:-LT3ORf0CMM3g4s-ZKgA:b-q2uk1p) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d64f2742554bb1b5fe0492ef2efe19f3239af9b/pkg/sql/opt/optbuilder/testdata/insert#L689)):*
> ```
> 
> # Allow NULL DEFAULT value to be inserted into not-null column (would fail at
> # runtime if there are any rows to update).
> ```

I don't understand this comment... isn't this going to insert one row with a null? Or is the point that this is an execution error, not a planning error?


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32774)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/opt/catalog.go, line 198 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LT3TuLf-MGpSOTlv76P:-LT3TuLf-MGpSOTlv76Q:bkbfpg5) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d64f2742554bb1b5fe0492ef2efe19f3239af9b/pkg/sql/opt/catalog.go#L198)):*
> ```Go
> // dropped from a table. Mutation columns require special handling by mutation
> // operators like Insert, Update, and Delete.
> type MutationColumn struct {
> ```

Wouldn't it be simpler to add `IsMutationColumn (ok, deleteOnly bool)` to `Column` and make this wrapper be private in the implementation? (we'd need to add a shell implementation of this to `ColumnDescriptor` but that seems fine)
___
*[pkg/sql/opt/catalog.go, line 209 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LT3ToIW6fLZ-XavXQiw:-LT3ToIW6fLZ-XavXQix:b-vcedcf) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d64f2742554bb1b5fe0492ef2efe19f3239af9b/pkg/sql/opt/catalog.go#L209)):*
> ```Go
> 
> // ColName is part of the opt.Column interface.
> func (c *MutationColumn) ColName() tree.Name {
> ```

[nit] `Column` could be embedded so we don't have to define these
___
*[pkg/sql/opt/ops/statement.opt, line 60 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSpp_iy1-9z2cp2fqh6:-LT3Ued8Brd3D37hb69v:b-en0j09) ([raw file](https://github.com/cockroachdb/cockroach/blob/382b1dcab6fffe9e1853435f23dd0a379b0aecb3/pkg/sql/opt/ops/statement.opt#L60)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

The repetition is a clever literary device that emphasizes the fact that not just one, but multiple columns may need to be updated at once.
</blockquote></details>

nit: I think you meant "The repetition is a clever is a clever literary device"


<!-- Sent from Reviewable.io -->

andy-kimball committed:
opt: Add experimental_optimizer_updates flag

When enabled, this flag uses the cost-based optimizer rather than the
heuristic planner to plan UPDATE statements. The CBO doesn't yet handle
multiple column families, so this flag avoids turning it on by default
so that performance in column family cases will not be affected for the
alpha release.

Release note (sql change): Added experimental_optimize_updates flag,
which uses the cost-based optimizer to plan UPDATE statements when set
to true.

Release note: None

andy-kimball reviewed (commented):
I also added a commit to this PR that only enables CBO UPDATE when the `experimental_optimizer_updates` session flag is set to `True`, in order to address concerns raised by +@jordanlewis that column family perf would otherwise regress due to missing "needed column" optimizations.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-:-LTDkulcAJIC1i44jxKd:bl0gtly)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/opt/catalog.go, line 198 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LT3TuLf-MGpSOTlv76P:-LT4Pzil-H9tCY2UYWgw:b-4u2mf6) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d64f2742554bb1b5fe0492ef2efe19f3239af9b/pkg/sql/opt/catalog.go#L198)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Wouldn't it be simpler to add `IsMutationColumn (ok, deleteOnly bool)` to `Column` and make this wrapper be private in the implementation? (we'd need to add a shell implementation of this to `ColumnDescriptor` but that seems fine)
</blockquote></details>

Hmm, maybe, but then if we add additional mutation fields (like `add` vs. `drop`), we'd need to add more return values to `IsMutationColumn`, which seems messy.
___
*[pkg/sql/opt/catalog.go, line 209 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LT3ToIW6fLZ-XavXQiw:-LT4Qf8q4jGFV-f1tzq_:b-fim5v0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d64f2742554bb1b5fe0492ef2efe19f3239af9b/pkg/sql/opt/catalog.go#L209)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] `Column` could be embedded so we don't have to define these
</blockquote></details>

Nice, I didn't know Go supported embedding an interface in a struct like that.
___
*[pkg/sql/opt/ops/statement.opt, line 60 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LSpp_iy1-9z2cp2fqh6:-LT4Qp_31T9k0S7_UhxR:b-wvr5va) ([raw file](https://github.com/cockroachdb/cockroach/blob/382b1dcab6fffe9e1853435f23dd0a379b0aecb3/pkg/sql/opt/ops/statement.opt#L60)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

nit: I think you meant "The repetition is a clever is a clever literary device"
</blockquote></details>

I'd hardly classify hardly classify your excellent suggestion as a nit. 
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 703 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LT3N7HE0vZviYTpB0Oj:-LT4RI3d0V_AGjKEcec1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d64f2742554bb1b5fe0492ef2efe19f3239af9b/pkg/sql/opt/optbuilder/mutation_builder.go#L703)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] These does -> These do
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/testdata/insert, line 730 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LT3O70a77FQHm2AKp85:-LT4qY9BBfXQKaWX3Xb3:b-n87xy6) ([raw file](https://github.com/cockroachdb/cockroach/blob/382b1dcab6fffe9e1853435f23dd0a379b0aecb3/pkg/sql/opt/optbuilder/testdata/insert#L730)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Why did you take out this test?
</blockquote></details>

It's no longer a compile-time error. It's now a runtime error. I added a new test above to ensure that it gets compiled correctly:
```
# Verify that there is no compile-time error when trying to insert a NULL
# DEFAULT value into a not-null column (it will fail at runtime).
build
INSERT INTO abcde (a) VALUES (DEFAULT)
----
```
I did this in response to Justin's comment on an earlier iteration, in which he pointed out that `UPDATE` only fails at runtime if trying to set a not-NULL column to NULL.
___
*[pkg/sql/opt/optbuilder/testdata/insert, line 689 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/32774#-LT3ORf0CMM3g4s-ZKg9:-LT4RRP8Dr7-13ihoenF:b-uhjc0t) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d64f2742554bb1b5fe0492ef2efe19f3239af9b/pkg/sql/opt/optbuilder/testdata/insert#L689)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

I don't understand this comment... isn't this going to insert one row with a null? Or is the point that this is an execution error, not a planning error?
</blockquote></details>

It's an execution error, not a planning error. I'll fix the comment.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
> Reviewed 1 of 86 files at r10, 3 of 10 files at r22, 27 of 30 files at r23, 5 of 20 files at r24, 2 of 3 files at r25, 25 of 26 files at r26, 22 of 23 files at r27, 11 of 11 files at r28.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32774)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

andy-kimball commented:
bors r+
